users/pipelines.py

```diff
@@ -25,9 +25,59 @@ def create_or_update_user(strategy, details, backend, user=None, *args, **kwargs
             last_name=details.get('last_name', ''),
             role=UserRole.objects.get(role='user')
         )
-        send_welcome_oauth_email_task.delay(
-            subject="Welcome!",
-            message="Thanks for signing up via OAuth",
-            recipient_list=[user.email]
-        )
     return {'user': user}
+
+def activate_verified_user(strategy, details, backend, user=None, *args, **kwargs):
+    """
+    Activate user only if email is verified. Critical security function.
+    """
+    if user and not user.is_active:
+        verified_email = False
+        response = kwargs.get('response', {})
+        
+        if backend.name == 'google-oauth2':
+            verified_email = response.get('email_verified', False)
+            
+        elif backend.name == 'github':
+            social = kwargs.get('social') or (
+                user.social_auth.filter(provider='github').first() if user else None
+                )        
+            if social and social.extra_data:
+                emails = social.extra_data.get('emails', []) 
+                user_email = user.email or details.get('email', '')
+                for email_data in emails:
+                    if (isinstance(email_data, dict) and 
+                        email_data.get('email') == user_email and 
+                        email_data.get('verified', False)):
+                        verified_email = True
+                        break
+            else:
+                verified_email = True     
```

Comment:
The else block for Github verification defaults to trusted if emails are missing, which could be insecure. Require explicit email verification for Github and handle missing emails defensively.
```
else:
    verified_email = False  # do not trust if emails are missing
```